Kernel/VirtIO/VirtIOConsole.h

```diff
@@ -38,10 +38,39 @@ class VirtIOConsole final : public CharacterDevice
     virtual String device_name() const override { return String::formatted("hvc{}", minor()); }
     virtual void handle_queue_update(u16 queue_index) override;
 
-    OwnPtr<Region> m_receive_region;
-    OwnPtr<Region> m_transmit_region;
+    class VMRingBuf {
+    public:
+        VMRingBuf(const StringView region_name, size_t capacity);
+
+        bool has_space() const { return num_used_bytes < capacity_in_bytes; }
+        bool copy_data_in(const UserOrKernelBuffer& buffer, size_t length, FlatPtr& start_of_copied_data, size_t& bytes_copied);
+        void reclaim_space(FlatPtr region_start, size_t chunk_size)
+        {
+            VERIFY(start_of_used_paddr() == region_start);
```

Comment:
you dont use region_start anywhere besides here, is it really needed?

Reply:
It exists entirely for verifying that the region we are reclaiming starts at the start of the ringbuffer, bad things can happen if this is not true. I'm actually having a small issue right now, where sometimes this verification fails. I think the device might not be required to return the buffers it was provided in order, I can't actually find anything in the spec that requires this. If this isn't required, it might not be possible to do this with a ringbuffer.

Reply:
the device is required to follow the descriptor order, but are you sure the descriptor order matches the ring buffer order?

Reply:
I dont see it there either, but if i remember correctly that's how it is implemented in qemu

Reply:
Hmm, I think there may be something else in the kernel that's causing it. It only occurs when I have VIRTIO_DEBUG turned off and am writing to the virtioconsole very fast (my test program is `yes 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa'`). I've noticed that when I run this in the terminal emulator (not piped to /dev/hvc0), the terminal emulator sometimes messes up rendering it (it prints out more 'a's than it should for one line every few seconds). I'm starting to believe that there might be some bug in the kernel that only occurs when under heavy load writing.

Reply:
I figured out what was causing the bug, it was a subtle race condition when accessing the ring buffer meta-data. Wrapping accesses to it in a lock fixes it.